junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/RunListenerAdapter.java

```diff
@@ -83,41 +80,27 @@ else if (testIdentifier.isTest()) {
 	}
 
 	private SimpleReportEntry createReportEntry(TestIdentifier testIdentifier, Optional<Throwable> throwable) {
-		Optional<String> className = getClassName(testIdentifier);
-		if (className.isPresent()) {
-			StackTraceWriter traceWriter = new PojoStackTraceWriter(className.get(),
-				getMethodName(testIdentifier).orElse(""), throwable.orElse(null));
-			return new SimpleReportEntry(className.get(), testIdentifier.getDisplayName(), traceWriter, null);
-		}
-		else {
-			return new SimpleReportEntry(parentDisplayName(testIdentifier), testIdentifier.getDisplayName(), null);
-		}
-	}
+		String className = getClassName(testIdentifier);
+
+		StackTraceWriter traceWriter = new PojoStackTraceWriter(className, testIdentifier.getLegacyReportingName(),
+			throwable.orElse(null));
+		return new SimpleReportEntry(className, testIdentifier.getDisplayName(), traceWriter, null);
 
-	private Optional<String> getClassName(TestIdentifier testIdentifier) {
-		TestSource testSource = testIdentifier.getSource().orElse(null);
-		if (testSource instanceof ClassSource) {
-			return Optional.of(((ClassSource) testSource).getJavaClass().getName());
-		}
-		if (testSource instanceof MethodSource) {
-			return Optional.of(((MethodSource) testSource).getClassName());
-		}
-		return Optional.empty();
 	}
 
-	private Optional<String> getMethodName(TestIdentifier testIdentifier) {
-		TestSource testSource = testIdentifier.getSource().orElse(null);
-		if (testSource instanceof MethodSource) {
-			return Optional.of(((MethodSource) testSource).getMethodName());
+	private String getClassName(TestIdentifier testIdentifier) {
+		if (testIdentifier.isContainer()) {
+			return testIdentifier.getLegacyReportingName();
 		}
-		return Optional.empty();
+
+		return parentDisplayName(testIdentifier);
 	}
```

Comment:
I think we should leave `getClassName()` and `getMethodName()` as is but _only_ pass them to `PojoStackTraceWriter` to "optimize" the stacktrace.